core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -303,66 +282,30 @@ protected void containerIsStarted(InspectContainerResponse containerInfo) {
      * @return the port on which to check if the container is ready
      */
     protected Integer getLivenessCheckPort() {
-        if (exposedPorts.size() > 0) {
-            return getMappedPort(exposedPorts.get(0));
-        } else if (portBindings.size() > 0) {
-            return PortBinding.parse(portBindings.get(0)).getBinding().getHostPort();
-        } else {
-            return null;
-        }
+        return getExposedPorts().stream()
+                .findFirst()
+                .map(this::getMappedPort)
+                .orElseGet(
+                        () -> getPortBindings().stream()
+                                .findFirst()
+                                .map(binding -> PortBinding.parse(binding).getBinding().getHostPort())
+                                .orElse(null)
+                );
     }
 
     private void applyConfiguration(CreateContainerCmd createCommand) {
-
-        // Set up exposed ports (where there are no host port bindings defined)
-        ExposedPort[] portArray = exposedPorts.stream()
-                .map(ExposedPort::new)
-                .toArray(ExposedPort[]::new);
-
-        createCommand.withExposedPorts(portArray);
-
-        // Set up exposed ports that need host port bindings
-        PortBinding[] portBindingArray = portBindings.stream()
-                .map(PortBinding::parse)
-                .toArray(PortBinding[]::new);
-
-        createCommand.withPortBindings(portBindingArray);
-
-        if (commandParts != null) {
-            createCommand.withCmd(commandParts);
-        }
-
-        String[] envArray = env.stream()
-                .toArray(String[]::new);
-        createCommand.withEnv(envArray);
-
-        Bind[] bindsArray = binds.stream()
-                .toArray(Bind[]::new);
-        createCommand.withBinds(bindsArray);
-
-        Link[] linksArray = linkedContainers.entrySet().stream()
-                .map(entry -> new Link(entry.getValue().getContainerName(), entry.getKey()))
-                .collect(Collectors.toList())
-                .toArray(new Link[linkedContainers.size()]);
-        createCommand.withLinks(linksArray);
-
         createCommand.withPublishAllPorts(true);
 
-        String[] extraHostsArray = extraHosts.stream()
-        		 .toArray(String[]::new);
-        createCommand.withExtraHosts(extraHostsArray);
+        getTraits().stream().sorted(TRAIT_COMPARATOR).forEach(trait -> trait.configure(self(), createCommand));
```

Comment:
allow traits to implement `Comparable` interface (sometimes some traits wants to force ordering, i.e. "apply default command if it wasn't applied before", lowest priority)